jenkins/opensearch-dashboards/distribution-build.jenkinsfile

```diff
@@ -872,44 +872,55 @@ pipeline {
+                            }
+                        }
                     }
-                    expression {
-                        params.BUILD_PLATFORM.contains('linux')
+                    options {
+                        timeout(time: 90, unit: 'MINUTES')
                     }
-                    expression {
-                        params.BUILD_DISTRIBUTION.contains('tar')
+                    agent {
+                        docker {
+                            label AGENT_LINUX_X64
+                            image dockerAgent.image
+                            args dockerAgent.args
+                            registryUrl 'https://public.ecr.aws/'
+                            alwaysPull true
+                        }
                     }
-                }
-            }
-            agent {
-                docker {
-                    label AGENT_LINUX_X64
-                    image dockerAgent.image
-                    args dockerAgent.args
-                    registryUrl 'https://public.ecr.aws/'
-                    alwaysPull true
-                }
-            }
-            steps {
-                script {
-                    echo "env.ARTIFACT_URL_LINUX_X64_TAR: ${env.ARTIFACT_URL_LINUX_X64_TAR}"
-                    echo "env.ARTIFACT_URL_LINUX_ARM64_TAR: ${env.ARTIFACT_URL_LINUX_ARM64_TAR}"
+                    steps {
+                        script {
+                            while(true) {
+                                if (env.ARTIFACT_URL_LINUX_X64_TAR != null && env.ARTIFACT_URL_LINUX_ARM64_TAR != null) {
+                                    echo "env.ARTIFACT_URL_LINUX_X64_TAR: ${env.ARTIFACT_URL_LINUX_X64_TAR}"
+                                    echo "env.ARTIFACT_URL_LINUX_ARM64_TAR: ${env.ARTIFACT_URL_LINUX_ARM64_TAR}"
 
-                    buildDockerImage(
-                        inputManifest: "manifests/${INPUT_MANIFEST}",
-                        buildNumber: "${BUILD_NUMBER}",
-                        buildOption: "${BUILD_DOCKER}",
-                        artifactUrlX64: env.ARTIFACT_URL_LINUX_X64_TAR,
-                        artifactUrlArm64: env.ARTIFACT_URL_LINUX_ARM64_TAR
-                    )
+                                    buildDockerImage(
+                                        inputManifest: "manifests/${INPUT_MANIFEST}",
+                                        buildNumber: "${BUILD_NUMBER}",
+                                        buildOption: "${BUILD_DOCKER}",
+                                        artifactUrlX64: env.ARTIFACT_URL_LINUX_X64_TAR,
+                                        artifactUrlArm64: env.ARTIFACT_URL_LINUX_ARM64_TAR
+                                    )
+                                    break
+                                } else {
+                                    echo "Waiting for x64 and arm64 tar builds to complete, sleeping 120 seconds..."
+                                    sleep(time: 120, unit: 'SECONDS')
```

Comment:
I had tried something like this https://github.com/gaiksaya/opensearch-build/blob/6d4745692422ba7c7c89a55db110a6789ab97e9c/jenkins/opensearch/test-dist.jenkinsfile#L65-L70

Can you check if this is a better option than that? It will automatically start the build as soon those variables are set. If not the timeout will take care of timing out.

Reply:
Both are the same, in this we don't need to set additional env variables as we can leverage the ones already available.

Reply:
Wont that start the build as soon as the variables are set instead of waiting for next 120 seconds? Also wondering, if both or one of the stages failed, we should skip building docker altogether. However,in this case, it will wait for 90 minutes before timing out. Looks like we cannot check stage status (only job status). Can we add some check for that scenario?

Reply:
Once the variables are set that means the artifacts are available in S3 and can be used immediately, I don't see an issue here. Also as per logic it will only trigger the docker build once both, x64 and arm64 artifacts are available else timeout eventually. 
Stage status check is not trivial and there is no straight-forward way to do the same, will need some hack.

Reply:
We shouldn't be building docker just for x64 or arm64, either build for both else skip.

Reply:
Right! My point being, the average build time for tarball x64 and arm64 is 20-30min. So if they fail, we would have to wait more 60min (since docker timeout is 90min) for docker stage to timeout eventually.

Reply:
Sure, may be you can PR it since you have the changes done in your local. I will cancel or you can edit this as well. 
I am okay with both.

Reply:
On a second look, `env.TAR_X64_SUCCESS = 'true'` would only be set if the groovy lib that executes the actual workflow returns successfully. In case of failure it will exit out of the job without setting this variable. 
So it is same as what I'm doing, in case if success/failure both the solutions work same. 
Please correct me if I'm wrong.

Reply:
How about setting `env.TAR_X64_SUCCESS = 'false'` in post failure stage? Or any value you see fit?

```
post {
  success {
      script {
          env.TAR_ARM64_SUCCESS = 'succeeded'
      }
  }
  failure {
    script {
      env.TAR_ARM64_SUCCESS = 'failed'
    }
  }
}
```

Check for both values in the docker build succeeded and failed and act accordingly. The main motive of this change is to lower the build time and immediately build docker when tarball is ready. Looks like in case of failure we would be waiting 90min for the job to complete.

Reply:
Synced up offline, looks like this would be too much of hack. We can see how the change in this PR goes and iterate on it later. Thanks!